Added ability to randomly kill pipelines to the constellation.

asajeffrey opened the pull request:
Added flags:
- `--random-pipeline-closure-probability`: probability of each event triggering a forced exit of a randomly chosen pipeline.
- `--random-pipeline-closure-seed`: seed to use for the RNG (to remove a source of intermittent failure).

These are designed for testing the hardness of the constellation.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10179)

<!-- Reviewable:end -->

Commit:
Added ability to randomly kill pipelines to the constellation.

asajeffrey commented:
Things that we need to make sure about this PR:
1. If `--random-pipeline-closure-probability` isn't set, the PR changes nothing.
2. There's no security implications, even if an attacker manages to get servo into zombie apocalypse mode.
I think I managed both of these, but you never know...

asajeffrey commented:
Argh, forgot to run test-tidy.

asajeffrey commented:
There is a constellation hardening metabug: issue #10124.

emilio commented:
Looks good to me!

I left a comment about the `close_random_pipeline` function, but I could tolerate landing it as-is if you think being clearer there is unnecessary, so feel free to r=me if that's the case :)

-S-awaiting-review +S-awaiting-answer 

---

Reviewed 4 of 6 files at r1, 1 of 2 files at r2.
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/compositing/constellation.rs, line 370 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10179#-KDiJ6j71JL_O1syCEef:-KDiJ6j71JL_O1syCEeg:-1273323318) ([raw file](https://github.com/servo/servo/blob/e661fa3bf51a83f51f8046f84c022ca3d21f5b56/components/compositing/constellation.rs#L370)):*
I'd include in this comment something like "if `--random-pipeline-closure-probability` was specified at startup" or similar, just to make it clear when scanning the constellation code (in order to not having to look ~1000 lines below).

Renaming `close_random_pipeline` to `maybe_close_random_pipeline`, or even changing the signature to `close_random_pipeline(&mut self, rng: &mut StdRng, probability: f32)`, so it would be obvious from here that this will do nothing unless `random_pipeline_closure` is set would help.

These are nits though, and I won't blame you for landing this as-is :P

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179#-:-KDiLEV9Edays_NAkB1l:-855846672)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/compositing/constellation.rs, line 370 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10179#-KDiJ6j71JL_O1syCEef:-KDiRwAr5HK14sI9IXg8:1179119842) ([raw file](https://github.com/servo/servo/blob/e661fa3bf51a83f51f8046f84c022ca3d21f5b56/components/compositing/constellation.rs#L370)):*
Added some comments, and renamed `close_random_pipeline` to `maybe_close_random_pipeline`. Thanks for the review!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179)*
<!-- Sent from Reviewable.io -->

mbrubeck commented on components/compositing/constellation.rs:
```diff
@@ -347,6 +352,11 @@ impl<LTF: LayoutThreadFactory, STF: ScriptThreadFactory> Constellation<LTF, STF>
                 document_states: HashMap::new(),
                 webrender_api_sender: state.webrender_api_sender,
+                random_pipeline_closure: opts::get().random_pipeline_closure_probability.map(|prob| {
+                    let seed = opts::get().random_pipeline_closure_seed.unwrap_or_else(random);
```
When it a random seed, perhaps it should log the seed.

asajeffrey replied:
Done. Should we log using `debug!` or something else?

mbrubeck replied:
`debug!` is compiled out in release builds of Servo, so use `info!` if you want to use this feature in release.

asajeffrey replied:
Will do.

emilio commented:
Looks good to me, r=me with that typo fixed and the commits squashed :)

-S-awaiting-answer -S-awaiting-answer +S-needs-code-changes +S-needs-squash 

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[components/compositing/constellation.rs, line 370 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10179#-KDiJ6j71JL_O1syCEef:-KDj6ZYxLQVdF4lqDyj4:450166850) ([raw file](https://github.com/servo/servo/blob/e661fa3bf51a83f51f8046f84c022ca3d21f5b56/components/compositing/constellation.rs#L370)):*
nit: `s/pipelne/pipeline`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179#-:-KDj6hqrGEl_fQIpk9K8:549934878)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Squashed, typo fixed.

---

Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/10179#-:-KDjt4HYvSDJt9LIkvYL:290882623)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
@bors-servo: r=emilio

bors-servo commented:
:pushpin: Commit 1ec222c has been approved by `emilio`

<!-- @bors-servo r=emilio 1ec222caa1e2798da4676f495ac53968dae8c54f -->

bors-servo commented:
:hourglass: Testing commit 1ec222c with merge fde06cc...

bors-servo commented:
:broken_heart: Test failed - [gonk](http://build.servo.org/builders/gonk/builds/2232)

emilio commented:
Need to update the other lockfiles.

asajeffrey commented:
Grr, I'll do this tomorrow.

asajeffrey commented:
OK, fixed Cargo.lock files, Let's see if this works...

@bors-servo try

bors-servo commented:
:hourglass: Trying commit 32c72f0 with merge 8a13b7b...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2166), [arm32](http://build.servo.org/builders/arm32/builds/95), [gonk](http://build.servo.org/builders/gonk/builds/2239), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2217), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2501), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/908), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2263), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2325), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.1196)

asajeffrey commented:
@bors-servo: r=emilio

bors-servo commented:
:pushpin: Commit 32c72f0 has been approved by `emilio`

<!-- @bors-servo r=emilio 32c72f0925140baa4ecf1a969d604f6c9506bb4c -->

bors-servo commented:
:hourglass: Testing commit 32c72f0 with merge dc0e541...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2502)

asajeffrey commented:
@bors-servo: retry
* #9818

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/2167), [arm32](http://build.servo.org/builders/arm32/builds/96), [gonk](http://build.servo.org/builders/gonk/builds/2240), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2218), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/909), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2264), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2326), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.1201) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/2502)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/2167), [arm32](http://build.servo.org/builders/arm32/builds/96), [gonk](http://build.servo.org/builders/gonk/builds/2240), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2218), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2503), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/909), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2264), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2326), [status-appveyor](https://ci.appveyor.com/project/servo/servo/build/1.0.1201)
